File: custinfo.scala
import org.apache.spark.sql.types._
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._

// need to read JSON; using the standard parser but it's deprecated!
import scala.util.parsing.json.JSON._
// read JSON as a string
val names_json_as_string = scala.io.Source.fromFile("data/data_names_.json").mkString
// convert this string to a map
val name_db:Map[String,String] = parseFull(names_json_as_string).get.asInstanceOf[Map[String, String]]

val titles_json_as_string = scala.io.Source.fromFile("data/data_salutations_.json").mkString
val title_db:Map[String,Map[String, List[String]]] = parseFull(titles_json_as_string).get.asInstanceOf[Map[String,Map[String, List[String]]]]

// hypocs json is like {"al": ["alison", "alexandria", ...

val hypocs_json_as_string = scala.io.Source.fromFile("data/data_hypocorisms_.json").mkString
val hypoc_db:Map[String,Map[String, List[String]]] = parseFull(hypocs_json_as_string).get.asInstanceOf[Map[String,Map[String, List[String]]]]

val grammg_json_as_string = scala.io.Source.fromFile("data/data_grammgender_.json").mkString
val grammg_db:Map[String,String] = parseFull(grammg_json_as_string).get.asInstanceOf[Map[String, String]]


def getGenderTitle(s: String): String = {

//	string s is presumably a title

	for (g <- List("m","f")){

		val all_salutations = title_db("common")(g) ++ title_db("uncommon")(g)

		if (all_salutations.contains(s)){return g}
			}
	return ""
		}

def getGenderName(s: String): Option[String] = {

//	s is a string possibly containing name

	val nameparts_incommon: Option[Set[String]] = Option(s.split("[-_\\s]").toSet.intersect(name_db.keySet | hypoc_db.keySet))

	nameparts_incommon match {

		case Some(Set[String]) => {val name_cand = nameparts_incommon.maxBy(_.length)
			if (name_db.contains(name_cand)){Some(name_db(name_cand))}
				else {Some("noname")}
			}
		case None => Some("result")
	}


	// if (nameparts_incommon.nonEmpty){
	
	// 		val name_cand = nameparts_incommon.maxBy(_.length)

	// 		println(name_cand)
	
	// 		// if candidate is name (and not hypoc)
	// 		if (name_db.contains(name_cand)){return name_db(name_cand)}
	// 		else {
	// 			// find what names corresp.to hypocorism and are in the name database
	// 			val unfolded_hypoc = hypoc_db(name_cand).toSet.intersect(name_db.keySet)

	// 			println("unfolded_hypoc=" + unfolded_hypoc)
				
	// 			if (unfolded_hypoc.nonEmpty){
	// 								val hyp_cand = unfolded_hypoc.maxBy(_.length)
	// 									return name_db(hyp_cand)}
	// 							}
	// 				}
				}

def getDomain(s: String): Option[String] = {

	// given a string s, extract the domain part if this string happens to be an email

	try {
		Some(s.split("@")(1))
	}
	catch {
		case e: Exception => None
	}
}

def isStudentOrStaff(s: String): String = {

	// given a string s and if it's an email address, is it a university student or staff?

	val staffPattern = "\\b[a-z-]+\\.*[a-z-]+\\@[a-z]+\\.?edu\\.*(au)*\\b".r
	val studentPattern = "\\b\\d*\\w+\\d+\\w*\\@[a-z]+\\.?edu\\.*(au)*\\b".r

	val staff_match = staffPattern.findFirstIn(s)
	val student_match = studentPattern.findFirstIn(s)

	var sos = "no"

	if (staff_match.getOrElse("no") != "no") {sos = "staff"}

	if (student_match.getOrElse("no") != "no") {sos = "student"}

	return sos


}


def isUni(s: String): String = {

	// given a string s, figure out which university exactly it's from (if any)

	val p = getDomain(s).getOrElse(" ")

	p match {
			case "adelaide.edu.au" => "university of adelaide"
			case "acu.edu.au" => "australian catholic university"
			case "anu.edu.au" => "australian national university"
			case "bond.edu.au" => "bond university"
			case "canberra.edu.au" => "university of canberra"
			case "cqu.edu.au" => "cquniversity"
			case "cdu.edu.au" => "charles darwin university"
			case "csu.edu.au" => "charles sturt university"
			case "curtin.edu.au" => "curtin university"
			case "deakin.edu.au" => "deakin university"
			case "ecu.edu.au" => "edith cowan university"
			case "federation.edu.au" => "federation university"
			case "flinders.edu.au" => "flinders university"
			case "griffith.edu.au" => "griffith university"
			case "jcu.edu.au" => "james cook university"
			case "latrobe.edu.au" => "la trobe university"
			case "mq.edu.au" => "macquarie university"
			case "unimelb.edu.au" => "university of melbourne"
			case "monash.edu" => "monash university"
			case "murdoch.edu.au" => "murdoch university"
			case "une.edu.au" => "university of new england"
			case "unsw.edu.au" => "university of new south wales"
			case "newcastle.edu.au" => "university of newcastle"
			case "nd.edu.au" => "university of notre dame"
			case "uq.edu.au" => "university of queensland"
			case "qut.edu.au" => "qut"
			case "rmit.edu.au" => "rmit"
			case "unisa.edu.au" => "university of south australia"
			case "scu.edu.au" => "southern cross university"
			case "usq.edu.au" => "university of southern queensland"
			case "usc.edu.au" => "usc"
			case "swin.edu.au" => "swinburne university"
			case "sydney.edu.au" => "university of sydney"
			case "utas.edu.au" => "university of tasmania"
			case "uts.edu.au" => "university of technology sydney"
			case "vu.edu.au" => "victoria university"
			case "uwa.edu.au" => "university of western australia"
			case "westernsydney.edu.au" => "western sydney university"
			case "uow.edu.au" => "university of wollongong"
			case _ => "none"
			}
		}

val getGenderTitleUDF = udf[String, String](getGenderTitle)

val getGenderNameUDF = udf[Option[String], String](getGenderName)

val isUniUDF = udf[String, String](isUni)

val isStudentOrStaffUDF = udf[String, String](isStudentOrStaff)

// in spark-shell Spark session is readily available as spark
val spark = SparkSession.builder.master("local").appName("test session").getOrCreate()

// set a smaller number of executors because this is running locally
spark.conf.set("spark.sql.shuffle.partitions", "4")

val df = spark.read.option("inferSchema", "true").option("header", "true").csv("data/sample_LotusCustomer.csv.gz")
			
val df1 = df.filter(df("CustomerID").isNotNull)
			.filter(df("CustomerListId") === 2)
			.filter(!df("EmailAddress").contains("ticketek"))

val df2 = df1.withColumn("EmailAddress_", when(df1("EmailAddress").contains("@"), lower(df1("EmailAddress"))).otherwise(""))

val df3 = df2.withColumn("University", isUniUDF(df2("EmailAddress_")))
			.withColumn("isStudentOrStaff", isStudentOrStaffUDF(df2("EmailAddress_")))
			.withColumn("Postcode", regexp_extract(df2("Postcode"),"\\b(([2-8]\\d{3})|([8-9]\\d{2}))\\b",0))
			.withColumn("Salutation", when(lower(regexp_replace(df2("Salutation"),"[^A-Za-z]",""))
											.isin(List("mr","ms","mrs","dr","mister","miss"):_*), lower(df("Salutation")))
											.otherwise(""))
val df4 = df3.withColumn("DateOfBirth", when(year(df3("DateOfBirth")) < 1918, lit(null)).otherwise(df3("DateOfBirth")))
			.withColumn("FirstName", ltrim(lower(regexp_replace(df3("FirstName"), "[-]"," "))))
			.withColumn("LastName", ltrim(lower(regexp_replace(regexp_replace(df3("LastName"),"['`]",""),"[-]"," "))))
			.withColumn("MobilePhone", regexp_extract(regexp_replace(df3("MobilePhone"),"\\s",""),"(\\+*(?:61)*|0*)(4\\d{8})",2))
			.withColumn("State", lower(df3("State")))
			.withColumn("City", lower(df3("City")))
			.withColumn("CountryName", lower(df3("CountryName")))
			.withColumn("Gender_Title", getGenderTitleUDF(df3("Salutation")))
			.withColumn("Gender_Name", getGenderNameUDF(df3("FirstName")))
			.select("CustomerID", "Salutation", "Gender_Title", "FirstName", "LastName", "DateOfBirth", 
						"CreatedDate", "ModifiedDate", "EmailAddress","University", "isStudentOrStaff", "State", "City","Postcode",
								"CountryName","MobilePhone", "HomePhone", "WorkPhone")
			.repartition(1)
			.write.option("header","true").mode("overwrite").option("compression", "gzip").csv("out")

File: custinfo.py
from pyspark.sql.types import *
from pyspark.sql import SparkSession
from pyspark.sql.functions import *
import re
import os
import json
import builtins

# --- load data files

# business emails: exact email to business type
bemails_db = json.load(open('data/business_email_dict.json'))
# business phone: phone number to business type
bephone_db = json.load(open('data/business_phone_dict.json'))
# education: domain to institution name
education_db = json.load(open('data/education_dict.json'))
# get useful files for gender detection
name_db, title_db, hypoc_db, grammg_db = [json.load(open(f,'r')) for f in [os.path.join(os.path.dirname(__file__),'data/data_names_.json'), 
																	 		os.path.join(os.path.dirname(__file__),'data/data_salutations_.json'), 
																	 		os.path.join(os.path.dirname(__file__),'data/data_hypocorisms_.json'),
																	 		os.path.join(os.path.dirname(__file__),'data/data_grammgender_.json')]]

# email domain parts likely favored by testers
testers_db = [line.strip() for line in open('data/tester_domains.txt').readlines() if line.strip()]

# role-based email prefixes
rolebased_db = [line.strip() for line in open('data/rolebased_email_prefixes.txt').readlines() if line.strip()]

# hotel domains
hotel_domains_db = [line.strip() for line in open('data/hotel_domains.txt').readlines() if line.strip()]

# hotel emails
hotel_emails_db = [line.strip() for line in open('data/hotel_emails.txt').readlines() if line.strip()]

# travel agent domains
ta_domains_db = [line.strip() for line in open('data/ta_domains.txt').readlines() if line.strip()]

# travel agent emails
ta_emails_db = [line.strip() for line in open('data/ta_emails.txt').readlines() if line.strip()]

def isTester(name, email_domain):

	if (not name) or (not email_domain):
		return None
		
	has_tester_domain = 0

	for t in testers_db:
		if t in str(email_domain):
			has_tester_domain += 1
			break

	has_tester_name = 1 if set(name.lower().split()) & set('test testing'.split()) else 0

	if has_tester_name and has_tester_domain:
		return 'yes'
	else:
		return 'no'


def getGenderTitle(s):
	"""
	string s is presumably a title
	"""
	for g in 'm f'.split():
		if s and (s.lower() in title_db['common'][g] + title_db['uncommon'][g]):
				return g

def getGenderName(s):
	"""
	s is a string possibly containing name
	"""
	if not s:
		return None

	# any name found straight away or hypocorisms? hypoc_db has hypocs as keys
	nameparts_incommon = {_ for w in s.split() for _ in w.split('-')} & (set(name_db) | set(hypoc_db))

	if nameparts_incommon:

		name_cand = builtins.max(nameparts_incommon, key=len)

		# if candidate is name (and not hypoc)
		if name_cand in name_db:
			return name_db[name_cand]
		else:
			# find what names corresp.to hypocorism and are in the name database
			unfolded_hypoc = set(hypoc_db[name_cand]) & set(name_db)

			if unfolded_hypoc:
				hyp_cand = builtins.max(unfolded_hypoc, key=len)
				return name_db[hyp_cand]

def getGenderEmail(s):

	if not s:
		return None
		
	# find any names in the email prefix; ONLY those separated by _, - or . count
	emailparts = set(q.strip() for w in s.split('_') 
										for v in w.split('.') 
											for q in v.split('-') if q.strip())

	names_in_email_ = emailparts & set(name_db)
	hypocs_in_email_ = emailparts & set(hypoc_db)
	grammgend_in_email_ = emailparts & set(grammg_db)

	if names_in_email_:

		_ = builtins.max(names_in_email_, key=len)

		if _ and (name_db[_] != 'u'):
			return name_db[_]
	
	if hypocs_in_email_:

		_ = builtins.max(hypocs_in_email_, key=len)
	
		if _ and (_ in name_db) and (name_db[_]!= 'u'):
			return name_db[longest_hyp]
	
	# last resort: grammatical gender
	# if grammgend_in_email_:

	# 	_ = builtins.max(grammgend_in_email_, key=len)
	
	# 	if _:
	# 		return grammg_db[_]


def isStudentOrStaff(s):

	# given a string s and if it's an email address, is it a university student or staff?

	staffPattern = re.compile(r"\b^[a-z]+[a-z-_\.]+\@[a-z\.]+\.?edu\.*(au)*\b")
	studentPattern = re.compile(r"\b^[a-z]+[a-z\.]*\d+.*\@[a-z\.]+\.?edu\.*(au)*\b")

	staff_match = re.match(staffPattern, s)
	student_match = re.match(studentPattern, s)

	if staff_match:
		return "staff"
	elif student_match:
		return "student"

def getBusiness(s):
	"""
	get business type by exact email address; to be applied on the column where the EMAIL ADDRESSES are 
	"""
	return bemails_db.get(s, None)

def getBusinessPhone(s):

	return bephone_db.get(s, None)

def isEducation(s):

	return education_db.get(s, None)


getGenderTitleUDF = udf(getGenderTitle, StringType())

getBusinessUDF = udf(getBusiness, StringType())

getBusinessPhoneUDF = udf(getBusinessPhone, StringType())

getGenderNameUDF = udf(getGenderName, StringType())

getGenderEmailUDF = udf(getGenderEmail, StringType())

isEducationUDF = udf(isEducation, StringType())

isStudentOrStaffUDF = udf(isStudentOrStaff, StringType())

isTesterUDF = udf(isTester, StringType())

# in pyspark Spark session is readily available as spark
spark = SparkSession.builder.master("local").appName("test session").getOrCreate()

# set a smaller number of executors because this is running locally
spark.conf.set("spark.sql.shuffle.partitions", "4")

df = spark.read.option("inferSchema", "true").option("header", "true").csv("data/sample_LotusCustomer.csv.gz")

# filter useless stuff out first
df1 = df.select("CustomerID", "Salutation", "FirstName", "LastName", "DateOfBirth", "EmailAddress", 
				"Postcode", "MobilePhone", "HomePhone", "WorkPhone", "CustomerListID") \
		.filter(df.CustomerID.isNotNull()) \
		.filter(df.CustomerListID == 2) 
		# .filter(~(df.EmailAddress.contains("@ticketek") | df.EmailAddress.contains("@teg-")))

# create a lower case trimmed email address 
df2 = df1.withColumn("EmailAddress", \
			when(df1.EmailAddress.contains("@"), rtrim(ltrim(lower(df1.EmailAddress)))).otherwise("")) \
		 .withColumn("Postcode", regexp_extract(df1.Postcode,"\\b(([2-8]\\d{3})|([8-9]\\d{2}))\\b",0)) \
		 .withColumn("Salutation", when(lower(regexp_replace(df1.Salutation,"[^A-Za-z]","")) \
		 								.isin(["mr","ms","mrs","dr","mister","miss"]), lower(df1.Salutation)) \
		 								.otherwise("")) \
		 .withColumn("DateOfBirth", when(year(df1.DateOfBirth) < 1918, lit(None)).otherwise(df1.DateOfBirth)) \
		 .withColumn("FirstName", rtrim(ltrim(lower(regexp_replace(df1.FirstName, "[-]"," "))))) \
		 .withColumn("LastName", rtrim(ltrim(lower(regexp_replace(regexp_replace(df1.LastName,"['`]",""),"[-]"," "))))) \
		 .withColumn("MobilePhone", regexp_extract(regexp_replace(df1.MobilePhone,"\\s",""),"(\\+*(?:61)*|0*)(4\\d{8})",2))

df2 = df2.withColumn("Age", floor(datediff(current_date(), "DateOfBirth")/365))

df3  = df2.withColumn("Gender_Title", getGenderTitleUDF(df2.Salutation)) \
		 	.withColumn("Gender_Name", getGenderNameUDF(df2.FirstName)) \
		 	.withColumn("Gender_Email", getGenderEmailUDF(df2.EmailAddress))

df3 = df3.withColumn("Gender", when(df3.Gender_Title.isNotNull(), 
										df3.Gender_Title).otherwise(when(df3.Gender_Name.isNotNull(), 
											df3.Gender_Name).otherwise(df3.Gender_Email))).drop("Gender_Title", "Gender_Name", "Gender_Email")


df3 = df3.withColumn("_tmp", \
					split(df3.EmailAddress, '@'))
df3 = df3.withColumn("EmailDomain", df3._tmp.getItem(1)).withColumn("EmailPrefix", df3._tmp.getItem(0)).drop("_tmp")
df3 = df3.withColumn("EmailType", when(df3.EmailPrefix.isin(rolebased_db), "role based"))

df3 = df3.withColumn("isTester", isTesterUDF(concat(df3.FirstName, lit(' '), df3.LastName), df3.EmailDomain))

df3 = df3.withColumn("Education", isEducationUDF(df3.EmailDomain)) \
			.withColumn("StudentOrStaff", isStudentOrStaffUDF(df3.EmailAddress)) \
			.withColumn('Business1', getBusinessUDF(df3.EmailAddress)) \
			 .withColumn('Business2', getBusinessPhoneUDF(df3.MobilePhone))

df3 = df3.withColumn("Business", when(df3.Business1.isNotNull(), df3.Business1).otherwise(df3.Business2)).drop("Business1", "Business2")

# travel agents; it's important that business type is 'travel agents' because that's what it is on YellowPages
df3 = df3.withColumn("isTA1", when(df3.EmailAddress.isin(ta_emails_db), "travel agents"))
df3 = df3.withColumn("isTA2", when(df3.EmailDomain.isin(ta_domains_db), "travel agents"))

df3 = df3.withColumn("Business", when(df3.isTA1.isNotNull(), 
										df3.isTA1).otherwise(when(df3.isTA2.isNotNull(), 
											df3.isTA2).otherwise(df3.Business))).drop("isTA1", "isTA2")

df3 = df3.withColumn("isHotel1", when(df3.EmailAddress.isin(hotel_emails_db), "hotels & accommodation"))
df3 = df3.withColumn("isHotel2", when(df3.EmailDomain.isin(hotel_domains_db), "hotels & accommodation"))

df3 = df3.withColumn("Business", when(df3.isHotel1.isNotNull(), 
										df3.isHotel1).otherwise(when(df3.isHotel2.isNotNull(), 
											df3.isHotel2).otherwise(df3.Business))).drop("isHotel1", "isHotel2")


df3.filter(df3.isTester.isNotNull()).select("CustomerID", "Salutation", "FirstName", "LastName", "Gender", "DateOfBirth", 
	"Age", "EmailAddress", "Business", "EmailType", "Education", "StudentOrStaff", "isTester").show()

df3.repartition(1).write.mode("overwrite").parquet("out")